Mathlib/Topology/Semicontinuous.lean

```diff
@@ -170,6 +172,61 @@ theorem LowerSemicontinuous.lowerSemicontinuousOn (h : LowerSemicontinuous f) (s
     LowerSemicontinuousOn f s := fun x _hx => h.lowerSemicontinuousWithinAt s x
 #align lower_semicontinuous.lower_semicontinuous_on LowerSemicontinuous.lowerSemicontinuousOn
 
+/-! #### Equivalent definitions -/
+
+
+section
+
+variable {γ : Type*} [CompleteLinearOrder γ] [DenselyOrdered γ]
+
+theorem lowerSemicontinuousAt_iff_le_liminf {f : α → γ} :
+    LowerSemicontinuousAt f x ↔ f x ≤ (liminf f <| 𝓝 x) := by
+  constructor
+  · intro hf; unfold LowerSemicontinuousAt at hf
+    contrapose! hf
+    obtain ⟨y,lty,ylt⟩ := exists_between hf; use y
+    exact ⟨ylt, fun h => not_le_of_lt lty
+      (Filter.le_liminf_of_le (by isBoundedDefault)
+        (Eventually.mono h (fun _ hx => le_of_lt hx)))⟩
+  exact fun hf y ylt => Filter.eventually_lt_of_lt_liminf (lt_of_lt_of_le ylt hf)
+
+theorem LowerSemicontinuousAt.le_liminf {f : α → γ} (hf : LowerSemicontinuousAt f x) :
+    f x ≤ (liminf f <| 𝓝 x) :=
+  lowerSemicontinuousAt_iff_le_liminf.mp hf
+
+theorem lowerSemicontinuous_iff_le_liminf {f : α → γ} :
+    LowerSemicontinuous f ↔ ∀ x, f x ≤ (liminf f <| 𝓝 x) := by
+  simp only [← lowerSemicontinuousAt_iff_le_liminf, LowerSemicontinuous]
+
+theorem LowerSemicontinuous.le_liminf {f : α → γ} (hf : LowerSemicontinuous f) :
+    ∀ x, f x ≤ (liminf f <| 𝓝 x) :=
+  lowerSemicontinuous_iff_le_liminf.mp hf
+
+theorem lowerSemicontinuousWithinAt_iff_le_liminf {f : α → γ} :
+    LowerSemicontinuousWithinAt f s x ↔ f x ≤ (liminf f <| 𝓝[s] x) := by
+  constructor
+  · intro hf; unfold LowerSemicontinuousWithinAt at hf
+    contrapose! hf
+    obtain ⟨y,lty,ylt⟩ := exists_between hf; use y
+    exact ⟨ylt, fun h => not_le_of_lt lty
+      (Filter.le_liminf_of_le (by isBoundedDefault)
+        (Eventually.mono h (fun _ hx => le_of_lt hx)))⟩
+  exact fun hf y ylt => Filter.eventually_lt_of_lt_liminf (lt_of_lt_of_le ylt hf)
```

Comment:
```suggestion
theorem lowerSemicontinuousWithinAt_iff_le_liminf {f : α → γ} :
    LowerSemicontinuousWithinAt f s x ↔ f x ≤ liminf f (𝓝[s] x) := by
  constructor
  · intro hf; unfold LowerSemicontinuousWithinAt at hf
    contrapose! hf
    obtain ⟨y, lty, ylt⟩ := exists_between hf; use y
    exact ⟨ylt, fun h => lty.not_le
      (le_liminf_of_le (by isBoundedDefault) (h.mono fun _ hx => le_of_lt hx))⟩
  exact fun hf y ylt => eventually_lt_of_lt_liminf (ylt.trans_le hf)
```
I cleaned up this proof a bit.

* Can you cleanup the statement of all other results in the same way? `f x ≤ (liminf f <| 𝓝[s] x)` is uglier than `f x ≤ liminf f (𝓝[s] x)`)
* Can you prove `lowerSemicontinuousAt_iff_le_liminf` from this version (and put this version first)? (use `lowerSemicontinuousWithinAt_univ_iff`)